pkg/operators/clustermanager/controllers/certrotationcontroller/certrotation_controller.go

```diff
@@ -0,0 +1,166 @@
+	clusterManagerLister operatorlister.ClusterManagerLister
+}
+
+func NewCertRotationController(
+	kubeClient kubernetes.Interface,
+	secretInformer corev1informers.SecretInformer,
+	configMapInformer corev1informers.ConfigMapInformer,
+	clusterManagerInformer operatorinformer.ClusterManagerInformer,
+	recorder events.Recorder,
+) factory.Controller {
+	signingRotation := SigningRotation{
+		Namespace:        helpers.ClusterManagerNamespace,
+		Name:             signerSecret,
+		SignerNamePrefix: signerNamePrefix,
+		Validity:         SigningCertValidity,
+		Lister:           secretInformer.Lister(),
+		Client:           kubeClient.CoreV1(),
+		EventRecorder:    recorder,
+	}
+	caBundleRotation := CABundleRotation{
+		Namespace:     helpers.ClusterManagerNamespace,
+		Name:          caBundleConfigmap,
+		Lister:        configMapInformer.Lister(),
+		Client:        kubeClient.CoreV1(),
+		EventRecorder: recorder,
+	}
+	targetRotations := []TargetRotation{
+		{
+			Namespace:     helpers.ClusterManagerNamespace,
+			Name:          helpers.RegistrationWebhookSecret,
+			Validity:      TargetCertValidity,
+			HostNames:     []string{fmt.Sprintf("%s.%s.svc", helpers.RegistrationWebhookService, helpers.ClusterManagerNamespace)},
+			Lister:        secretInformer.Lister(),
+			Client:        kubeClient.CoreV1(),
+			EventRecorder: recorder,
+		},
+		{
+			Namespace:     helpers.ClusterManagerNamespace,
+			Name:          helpers.WorkWebhookSecret,
+			Validity:      TargetCertValidity,
+			HostNames:     []string{fmt.Sprintf("%s.%s.svc", helpers.WorkWebhookService, helpers.ClusterManagerNamespace)},
+			Lister:        secretInformer.Lister(),
+			Client:        kubeClient.CoreV1(),
+			EventRecorder: recorder,
+		},
+	}
+
+	c := &CertRotationController{
+		SigningRotation:      signingRotation,
+		CABundleRotation:     caBundleRotation,
+		TargetRotations:      targetRotations,
+		kubeClient:           kubeClient,
+		clusterManagerLister: clusterManagerInformer.Lister(),
+	}
+	return factory.New().
+		ResyncEvery(ResyncInterval).
+		WithSync(c.Sync).
+		WithInformers(
+			secretInformer.Informer(),
+			configMapInformer.Informer(),
+			clusterManagerInformer.Informer(),
+		).
+		ToController("CertRotationController", recorder)
+}
+
+func (c CertRotationController) Sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	// do nothing if there is no cluster manager
+	clustermanagers, err := c.clusterManagerLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+	if len(clustermanagers) == 0 {
+		klog.V(4).Infof("No ClusterManager found")
+		return nil
+	}
+
+	klog.Infof("Reconciling ClusterManager %q", clustermanagers[0].Name)
+	// do nothing if the cluster manager is deleting
+	if !clustermanagers[0].DeletionTimestamp.IsZero() {
+		return nil
+	}
+
+	// create namespace for cluster manager if necessary
+	_, err = c.kubeClient.CoreV1().Namespaces().Get(ctx, helpers.ClusterManagerNamespace, metav1.GetOptions{})
+	switch {
```

Comment:
how about we just  return err here? ns will created finally

Reply:
I also updated clusterManagerController. It will not apply static resources, including the namespace, until the ca bundle configmap is created. So I have to create ns if it does not exist yet.

Reply:
i would apply  everything all  the time even  without a certain configmap. You finally get  into a steady  state, right? you may want to minimize coordination between two controllers.

Reply:
> i would apply everything all the time even without a certain configmap. You finally get into a steady state, right? you may want to minimize coordination between two controllers.

It makes sense.  Updated as suggested.